File: ts/src/xpath/tokenizer.ts
export type XPathTokenKind =
  | "identifier"
  | "number"
  | "string"
  | "operator"
  | "paren_open"
  | "paren_close"
  | "bracket_open"
  | "bracket_close"
  | "dot"
  | "dotdot"
  | "slash"
  | "comma"
  | "eof";

export type XPathToken = {
  kind: XPathTokenKind;
  value: string;
  position: number;
};

const OPERATORS = ["<=", ">=", "!=", "==", "//", "=", "<", ">", "+", "-", "*", "/"];

export function tokenizeXPath(expr: string): XPathToken[] {
  const tokens: XPathToken[] = [];
  let position = 0;

  const add = (kind: XPathTokenKind, value: string, at = position): void => {
    tokens.push({ kind, value, position: at });
  };

  while (position < expr.length) {
    const ch = expr[position];

    if (/\s/.test(ch)) {
      position += 1;
      continue;
    }

    if (ch === "\"" || ch === "'") {
      const quote = ch;
      const start = position;
      position += 1;
      let value = "";
      while (position < expr.length) {
        const c = expr[position];
        if (c === quote && expr[position - 1] !== "\\") {
          position += 1;
          break;
        }
        value += c;
        position += 1;
      }
      add("string", value, start);
      continue;
    }

    if (/\d/.test(ch) || (ch === "-" && /\d/.test(expr[position + 1] ?? ""))) {
      const start = position;
      let value = "";
      if (expr[position] === "-") {
        value += "-";
        position += 1;
      }
      while (position < expr.length && /\d/.test(expr[position])) {
        value += expr[position];
        position += 1;
      }
      if (expr[position] === "." && /\d/.test(expr[position + 1] ?? "")) {
        value += ".";
        position += 1;
        while (position < expr.length && /\d/.test(expr[position])) {
          value += expr[position];
          position += 1;
        }
      }
      add("number", value, start);
      continue;
    }

    if (/[A-Za-z_]/.test(ch)) {
      const start = position;
      let value = "";
      while (position < expr.length && /[A-Za-z0-9_:\-]/.test(expr[position])) {
        value += expr[position];
        position += 1;
      }
      add("identifier", value, start);
      continue;
    }

    if (ch === "/") {
      if (expr[position + 1] === "/") {
        add("operator", "//");
        position += 2;
      } else {
        add("slash", "/");
        position += 1;
      }
      continue;
    }

    if (ch === ".") {
      if (expr[position + 1] === ".") {
        add("dotdot", "..");
        position += 2;
      } else {
        add("dot", ".");
        position += 1;
      }
      continue;
    }

    if (ch === "(") {
      add("paren_open", "(");
      position += 1;
      continue;
    }
    if (ch === ")") {
      add("paren_close", ")");
      position += 1;
      continue;
    }
    if (ch === "[") {
      add("bracket_open", "[");
      position += 1;
      continue;
    }
    if (ch === "]") {
      add("bracket_close", "]");
      position += 1;
      continue;
    }
    if (ch === ",") {
      add("comma", ",");
      position += 1;
      continue;
    }

    const remaining = expr.slice(position);
    const op = OPERATORS.find((candidate) => remaining.startsWith(candidate));
    if (op) {
      add("operator", op);
      position += op.length;
      continue;
    }

    position += 1;
  }

  add("eof", "", position);
  return tokens;
}

File: src/xyang/xpath/tokenizer.py
"""
XPath expression tokenizer for xpath.
"""

from typing import List, Optional, Set

from .tokens import Token, TokenType

QUOTE_CHARS: Set[str] = {'"', "'"}
OPERATOR_CHARS: Set[str] = {'=', '<', '>', '!', '+', '-', '*', '/'}


class XPathTokenizer:
    """Tokenizer for XPath expressions."""

    OPERATORS = [
        '<=', '>=', '!=', '==', '//',
        '=', '<', '>', '+', '-', '*', '/',
    ]

    KEYWORDS = ['or', 'and', 'not', 'true', 'false', 'current']

    def __init__(self, expression: str):
        self.expression = expression
        self.position = 0
        self.tokens: List[Token] = []

    def tokenize(self) -> List[Token]:
        """Tokenize the expression."""
        self.position = 0
        self.tokens = []

        while self.position < len(self.expression):
            if self._skip_whitespace():
                continue

            if self.position >= len(self.expression):
                break

            char = self.expression[self.position]

            if char in QUOTE_CHARS:
                self._tokenize_string()
            elif char.isdigit() or (char == '-' and self._is_next_digit()):
                self._tokenize_number()
            elif char.isalpha() or char == '_':
                self._tokenize_identifier()
            elif char == '/':
                self._tokenize_slash()
            elif char == '.':
                self._tokenize_dot()
            elif char in OPERATOR_CHARS:
                self._tokenize_operator()
            elif char == '(':
                self._add_simple_token(TokenType.PAREN_OPEN, '(', 1)
            elif char == ')':
                self._add_simple_token(TokenType.PAREN_CLOSE, ')', 1)
            elif char == '[':
                self._add_simple_token(TokenType.BRACKET_OPEN, '[', 1)
            elif char == ']':
                self._add_simple_token(TokenType.BRACKET_CLOSE, ']', 1)
            elif char == ',':
                self._add_simple_token(TokenType.COMMA, ',', 1)
            else:
                self.position += 1

        self.tokens.append(Token(TokenType.EOF, '', self.position))
        return self.tokens

    def _add_simple_token(self, token_type: TokenType, value: str, advance: int):
        self.tokens.append(Token(token_type, value, self.position))
        self.position += advance

    def _skip_whitespace(self) -> bool:
        if self.position < len(self.expression) and self.expression[self.position].isspace():
            while self.position < len(self.expression) and self.expression[self.position].isspace():
                self.position += 1
            return True
        return False

    def _is_next_digit(self) -> bool:
        return (self.position + 1 < len(self.expression) and
                self.expression[self.position + 1].isdigit())

    def _peek_char(self, offset: int = 0) -> Optional[str]:
        pos = self.position + offset
        return self.expression[pos] if 0 <= pos < len(self.expression) else None

    def _tokenize_string(self):
        quote_char = self.expression[self.position]
        start_pos = self.position
        self.position += 1

        value = []
        while self.position < len(self.expression):
            char = self.expression[self.position]
            if char == quote_char:
                if self.position > 0 and self.expression[self.position - 1] == '\\':
                    value[-1] = quote_char
                else:
                    self.position += 1
                    break
            value.append(char)
            self.position += 1

        self.tokens.append(Token(TokenType.STRING, ''.join(value), start_pos))

    def _tokenize_number(self):
        start_pos = self.position
        value = []

        if self.expression[self.position] == '-':
            value.append('-')
            self.position += 1

        while self.position < len(self.expression) and self.expression[self.position].isdigit():
            value.append(self.expression[self.position])
            self.position += 1

        if (self._peek_char() == '.' and self._peek_char(1) and
            self._peek_char(1).isdigit()):
            value.append('.')
            self.position += 1
            while self.position < len(self.expression) and self.expression[self.position].isdigit():
                value.append(self.expression[self.position])
                self.position += 1

        self.tokens.append(Token(TokenType.NUMBER, ''.join(value), start_pos))

    def _tokenize_identifier(self):
        start_pos = self.position
        value = []

        while self.position < len(self.expression):
            char = self.expression[self.position]
            if char.isalnum() or char in ('_', '-', ':'):
                value.append(char)
                self.position += 1
            else:
                break

        identifier = ''.join(value)
        self.tokens.append(Token(TokenType.IDENTIFIER, identifier, start_pos))

    def _tokenize_operator(self):
        start_pos = self.position
        remaining = self.expression[self.position:]

        for op in self.OPERATORS:
            if remaining.startswith(op):
                self.tokens.append(Token(TokenType.OPERATOR, op, start_pos))
                self.position += len(op)
                return

        char = self.expression[self.position]
        self.tokens.append(Token(TokenType.OPERATOR, char, start_pos))
        self.position += 1

    def _tokenize_dot(self):
        start_pos = self.position

        if self._peek_char(1) == '.':
            self.tokens.append(Token(TokenType.DOTDOT, '..', start_pos))
            self.position += 2
        else:
            self.tokens.append(Token(TokenType.DOT, '.', start_pos))
            self.position += 1

    def _tokenize_slash(self):
        start_pos = self.position

        if self._peek_char(1) == '/':
            self.tokens.append(Token(TokenType.OPERATOR, '//', start_pos))
            self.position += 2
        else:
            self.tokens.append(Token(TokenType.SLASH, '/', start_pos))
            self.position += 1
